### PR TITLE
Alpha beta pruning

### PR DESCRIPTION
**Pull request by piuswalter**

### CODE REVIEW

File: src/Exercise07AI.ipynb [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+    "\n",
+    "Beim Progressive Deepening werden die Züge bis zu einer gegebenen Tiefe `DEPTH` berechnet. Dabei wird bei Tiefe `1` begonnen um eine grobe Wertung der bis dahin möglichen Züge zu erhalten. Beim nächsten Aufruf von Tiefe `2` kann dann die vorherige Evaluierung genutzt werden, um die Züge anhand diesem Maß zu sortieren und das Pruning so signifikant zu erhöhen.\n",
+    "\n",
+    "Die Funktion `minimax` wird um eine mit [heapq](https://docs.python.org/3/library/heapq.html) implementierte Warteschlange `moves` erweitert. Dabei handelt es sich um eine Liste, in der die evaluierten Züge absteigend ihrer Bewertung iterativ aufgenommen werden. Durch die Funktionsweise der Sortierung von `heapq` wird ein Tupel der Form\n",
+    "\n",
+    "$$ key = (evaluation, i, move) $$\n",
+    "\n",
+    "erstellt, welches an zweiter Position $i$ hinzufügt, das sicher verschieden ist, um Python vom Werfen einer Exception bei der Sortierung von $move$ abzuhalten. Die Evaluieren der Züge wird in zwei Teile aufgeteilt:\n",
```
Comment (p-fruck) [inline src/Exercise07AI.ipynb:77]:
```suggestion
    "erstellt, welches an zweiter Position $i$ hinzufügt, das sicher verschieden ist, um Python vom Werfen einer Exception bei der Sortierung von $move$ abzuhalten. Das Evaluieren der Züge wird in zwei Teile aufgeteilt:\n",
```

File: src/Exercise07AI.ipynb [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+    "\n",
+    "1. Die in `board.legal_moves` verfügbaren Züge werden iterativ gegen den `local_cache` geprüft. Ist der Zug bereits mit einer Tiefe von `depth - 2` verfügbar, wird diese Evaluation genutzt, alternativ wird $i$ zurückgegeben, mit dem der Zug in der Warteschlange einsortiert wird.\n",
```
Comment (p-fruck) [inline src/Exercise07AI.ipynb:79]:
```suggestion
    "1. Die in `board.legal_moves` verfügbaren Züge werden iterativ gegen den `local_cache` geprüft. Ist der Zug bereits mit einer Tiefe von `depth - 2` verfügbar, wird diese Evaluation genutzt, alternativ wird $i$ zurückgegeben, womit der Zug in der Warteschlange einsortiert wird.\n",
```

File: scripts/server_setup.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+# Enable 4 GiB swap space
+if ! [ -f /mnt/swapfile ]; then
+  sudo fallocate --length 4GiB /mnt/swapfile
+  sudo chmod 600 /mnt/swapfile
+  sudo mkswap /mnt/swapfile
+  sudo swapon /mnt/swapfile
+  sudo sh -c 'echo "/mnt/swapfile swap swap defaults 0 0" >> /etc/fstab'
+  sudo sh -c 'echo "vm.swappiness=10" >> /etc/sysctl.conf'
+fi
```
Comment (p-fruck) [inline scripts/server_setup.sh:28]: Use `echo <content> | sudo tee -a <file>` to append data to a file without write privileges.
A fail-safe implementation which does not bloat the given files after repeated execution would look like this:

```suggestion
function replace_in_file() {
  if [ -z "$3" ]; then
    echo "Usage: line_in_file <file> <regex_start> <value>"
    exit 1
  fi
  if [ -z "$(grep ^$2 $1)" ]; then
    # append line to given file
    echo "$3" | sudo tee -a "$1"
  else
    # line already exists in file, replace it
    sudo sed -i "s|$2.*|$3|g" "$1"
  fi
}

# Enable 4 GiB swap space
if ! [ -f /mnt/swapfile ]; then
  sudo fallocate --length 4GiB /mnt/swapfile
  sudo chmod 600 /mnt/swapfile
  sudo mkswap /mnt/swapfile
  sudo swapon /mnt/swapfile
  replace_in_file /etc/fstab "/mnt/swapfile"  "/mnt/swapfile swap swap defaults 0 0"
  replace_in_file /etc/sysctl.conf "vm\.swappiness=" "vm.swappiness=10"
fi
```

File: src/Exercise07AI.ipynb [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+    "        self.local_cache: dict[tuple, int] = {}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Progressive Deepening\n",
+    "\n",
+    "Da das in `Exercise05AI` implementierte Alpha-Beta-Pruning am effizientesten ist, wenn zuerst die besten Züge untersucht werden, wird diese Klasse um eine Variable `local_cache` erweitert. Initial ist `local_cache` ein leeres Dictionary, welches als Schlüssel ein Tupel annimmt und als Wert eine Ganzzahl zurückgibt. Das Tupel besteht dabei aus Tiefe und Zug, die dadurch resultierende Bewertung wird als Value in diesem Dictionary gespeichert.\n",
```
Comment (antonplagemann) [inline src/Exercise07AI.ipynb:69]:
```suggestion
    "Da das in `Exercise05AI` implementierte Alpha-Beta-Pruning am effizientesten ist, wenn zuerst die besten Züge untersucht werden, wird nun die Liste der mögliche Züge vorab nach der bisher bekannten Evaluierung jedes Zuges sortiert. Hierzu wurde die Klasse um eine Variable `local_cache` erweitert, welche bisherige Evaluierungen speichert. Initial ist `local_cache` ein leeres Dictionary, welches als Schlüssel ein Tupel annimmt und als Wert eine Ganzzahl zurückgibt. Das Tupel besteht dabei aus der Tiefe und einem Zug. Die zugehörige Evaluierung dieses Zuges mit der gegebenen Tiefe wird als Value in diesem Dictionary gespeichert.\n",
```

File: src/Exercise07AI.ipynb [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+    "\n",
+    "Beim Progressive Deepening werden die Züge bis zu einer gegebenen Tiefe `DEPTH` berechnet. Dabei wird bei Tiefe `1` begonnen um eine grobe Wertung der bis dahin möglichen Züge zu erhalten. Beim nächsten Aufruf von Tiefe `2` kann dann die vorherige Evaluierung genutzt werden, um die Züge anhand diesem Maß zu sortieren und das Pruning so signifikant zu erhöhen.\n",
```
Comment (antonplagemann) [inline src/Exercise07AI.ipynb:71]:
```suggestion
    "Beim Progressive Deepening werden die Züge iterativ bis zu einer gegebenen Tiefe `DEPTH` berechnet. Dabei wird mit der Tiefe `1` begonnen um eine grobe Wertung der bis dahin möglichen Züge zu erhalten. Beim nächsten Aufruf mit Tiefe `2` kann dann die vorherige Evaluierung genutzt werden, um die Züge anhand diesem Maß zu sortieren und somit das Pruning signifikant zu erhöhen.\n",
```

File: src/Exercise07AI.ipynb [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+    "\n",
+    "Die Funktion `minimax` wird um eine mit [heapq](https://docs.python.org/3/library/heapq.html) implementierte Warteschlange `moves` erweitert. Dabei handelt es sich um eine Liste, in der die evaluierten Züge absteigend ihrer Bewertung iterativ aufgenommen werden. Durch die Funktionsweise der Sortierung von `heapq` wird ein Tripel der Form\n",
+    "\n",
+    "$$ key = (evaluation, i, move) $$\n",
+    "\n",
+    "erstellt, welches an zweiter Position $i$ hinzufügt, das sicher verschieden ist, um Python vom Werfen einer Exception bei der Sortierung von $move$ abzuhalten. Das Evaluieren der Züge wird in zwei Teile aufgeteilt:\n",
```
Comment (antonplagemann) [inline src/Exercise07AI.ipynb:77]:
```suggestion
    "Die Funktion `minimax` wird um eine mit [heapq](https://docs.python.org/3/library/heapq.html) implementierte Prioritäts-Warteschlange `moves` erweitert, welche alle verfügbaren Züge speichert. Als Priorität wird hierbei die Evaluierung des Zuges mit geringerer Tiefe verwendet. Falls keine Evaluierung vorhanden ist, wird eine fortlaufende Zugnummer als Priorität verwendet. Bedingt durch die Funktionsweise der Sortierung von `heapq` wird nun ein Tripel der Form\n",
    "\n",
    "$$ key = (-evaluation, i, move) $$\n",
    "\n",
    "erstellt. Hierbei wird $evaluation$ invertiert, da bei `heapq` der niedrigste Wert die höchste Priorität hat. Zusätzlich wird an zweiter Position eine fortlaufende Zugnummer $i$ hinzufügt. Dies ist notwendig, da die Evaluierung für zwei Züge gleich sein kann und `heapq` in diesem Fall die nächsten Elemente des Tupels vergleicht. Für $move$ ist das nicht möglich, daher wird zuvor $i$ eingefügt um diesen Vergleich (und die ansonsten entstehende `Exception`) zu verhindern. Die Berechnung der neuen Evaluierung erfolgt nun in zwei Schritten:\n",
```

File: src/Exercise07AI.ipynb [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+    "\n",
+    "1. Die in `board.legal_moves` verfügbaren Züge werden iterativ gegen den `local_cache` geprüft. Ist der Zug bereits mit einer Tiefe von `depth - 2` verfügbar, wird diese Evaluation genutzt, alternativ wird $i$ zurückgegeben, womit der Zug in der Warteschlange einsortiert wird.\n",
```
Comment (antonplagemann) [inline src/Exercise07AI.ipynb:79]:
```suggestion
    "1. Die in `board.legal_moves` verfügbaren Züge werden iterativ gegen den `local_cache` geprüft. Wenn die Evaluierung des Zuges bereits mit einer Tiefe von `depth - 2` verfügbar ist, wird diese genutzt, ansonsten wird die Zugnummer $i$ verwendet. Anschließend wird der Zug mit dem erhaltenen Wert als Priorität in die Warteschlange einsortiert.\n",
```

File: src/Exercise07AI.ipynb [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+    "2. An dieser Stelle sind alle möglichen Züge bereits in der Liste `moves` enthalten und können mithilfe von `heapq.heappop(moves)[2]` extrahiert werden. Nach der Evaluierung durch den Minimax-Alogrithmus wird das berechnete Ergebnis im `local_cache` gespeichert."
```
Comment (antonplagemann) [inline src/Exercise07AI.ipynb:80]:
```suggestion
    "2. An dieser Stelle sind alle möglichen Züge bereits in der Liste `moves` enthalten und der Zug mit der bisher besten Evaluierung kann mithilfe von `heapq.heappop(moves)[2]` extrahiert werden. Nach der Berechnung durch den Minimax-Alogrithmus wird das neue Ergebnis im `local_cache` gespeichert."
```

File: src/Exercise07LiteAI.ipynb [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+    "from Exercise06AI import Exercise06AI"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Aufgabe 07Lite: Minimax mit Alpha-Beta-Pruning und Progressive Deepening OHNE Memoisierung\n",
+    "\n",
+    "Dieses Notebook ist eine Kopie der `Exercise07AI` ohne Memoisierung."
```
Comment (antonplagemann) [inline src/Exercise07LiteAI.ipynb:44]:
```suggestion
    "Dieses Notebook ist eine Kopie der `Exercise07AI` ohne Memoisierung. Durch diese Modifizierung sinkt der verwendete Arbeitsspeicher erheblich und die Berechnungsdauer erhöht sich leicht."
```

File: src/Exercise07LiteAI.ipynb [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+    "        self.local_cache: dict[tuple, int] = {}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Entfernung der Memoisierung\n",
+    "\n",
+    "Diese Klasse ist bis auf das Entfernen der Memoisierung identisch zur Klasse `Exercise07AI` mit dem Vorteil des geringeren Speicherbedarfs."
```
Comment (antonplagemann) [inline src/Exercise07LiteAI.ipynb:69]:
```suggestion
    "Diese Klasse ist bis auf das Entfernen der Memoisierung identisch zur Klasse `Exercise07AI`."
```

File: src/Exercise07LiteAI.ipynb [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class Exercise07LiteAI(Exercise07LiteAI): # type: ignore\n",
+    "    \n",
+    "    def get_next_middle_game_move(self, board: chess.Board) -> chess.Move | None:\n",
```
Comment (antonplagemann) [inline src/Exercise07LiteAI.ipynb:152]: Can we remove this code? Should be identical to `Exercise07AI`.

**Review by antonplagemann — APPROVED** [review]

**Review by p-fruck — APPROVED** [review]

**Pull request merged**